plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,545 @@
+} WriteContext;
+
+static gboolean
+fu_mbim_qdu_updater_reboot_timeout (gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_mbim_qdu_updater_file_write_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	MbimMessage *request  = NULL;
+	MbimMessage *response = NULL;
+	g_autoptr(GError)      error = NULL;
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_write_response_parse (
+							 response,
+							 &error)) {
+		g_debug ("[%s][%d]: error: couldn't parse response message: %s\n", __FUNCTION__, __LINE__, error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	fu_device_set_progress_full (FU_DEVICE (ctx->device), (gsize) ctx->chunk_sent, (gsize) ctx->chunks->len);
+	ctx->chunk_sent++;
+	if (ctx->chunk_sent < ctx->chunks->len) {
+		FuChunk *chk = g_ptr_array_index (ctx->chunks, ctx->chunk_sent);
+		request = (mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), &error));
+		mbim_device_command (ctx->mbim_device,
+				     request,
+				     20,
+				     ctx->cancellable,
+				     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+				     ctx);
+		mbim_message_unref (request);
+	} else {
+		fu_device_set_progress (FU_DEVICE (ctx->device), 100);
+		g_debug ("Done!");
+		fu_device_set_status (FU_DEVICE (ctx->device), FWUPD_STATUS_DEVICE_RESTART);
+		g_warn_if_fail (ctx->timeout_id == 0);
+		/* device will auto reboot right after update finish */
+		ctx->timeout_id = g_timeout_add_seconds (30, fu_mbim_qdu_updater_reboot_timeout, ctx);
+	}
+}
+
+static void
+fu_mbim_qdu_updater_file_open_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	MbimMessage *request = NULL;
+	MbimMessage *response = NULL;
+	g_autoptr(GError)      error = NULL;
+	WriteContext *ctx = user_data;
+	guint32  out_max_transfer_size;
+	guint32  out_max_window_size;
+	guint32  total_sending_numbers = 0;
+	FuChunk *chk = NULL;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_open_response_parse (
+							response,
+							&out_max_transfer_size,
+							&out_max_window_size,
+							&error)) {
+		g_debug ("error: couldn't parse response message: %s\n", error->message);
+		g_debug ("out_max_transfer_size: %d\n", (int)out_max_transfer_size);
```

Comment:
You should only attempt to use the contents of the out parameters if mbim_message_qdu_file_open_response_parse succeeds. If the method fails, you can only read the error, and you must ignore all the other fields.